src/primitives/transaction.h

```diff
@@ -616,21 +619,146 @@ class CTxCertifierLockOut : public CTxCrosschainOut
     }
 };
 
+// forward declarations
+class CValidationState;
+class CTxMemPool;
+class CCoinsViewCache;
+class CChain;
+class CBlock;
+class CBlockTemplate;
+class CScriptCheck;
+class CBlockUndo;
+class CTxUndo;
+class UniValue;
+
+namespace Sidechain { class ScCoinsViewCache; }
+
+
+// abstract interface for CTransaction and CScCertificate
+class CTransactionBase
+{
+protected:
+    /** Memory only. */
+    const uint256 hash;
+
+    virtual void UpdateHash() const = 0;
+
+public:
+    virtual bool TryPushToMempool(bool fLimitFree, bool fRejectAbsurdFee) = 0;
+    const int32_t nVersion;
+    const std::vector<CTxOut> vout;
+
+    CTransactionBase();
+    CTransactionBase& operator=(const CTransactionBase& tx);
+    CTransactionBase(const CTransactionBase& tx);
+    virtual ~CTransactionBase() {};
+
+    template <typename Stream>
+    CTransactionBase(deserialize_type, Stream& s) : CTransactionBase(CMutableTransactionBase(deserialize, s)) {}
+
+    const uint256& GetHash() const {
+        return hash;
+    }
+
+    bool IsScVersion() const
+    {
+        // so far just one version
+        return (nVersion == SC_TX_BASE_VERSION);
+    }
+
+    friend bool operator==(const CTransactionBase& a, const CTransactionBase& b)
+    {
+        return a.hash == b.hash;
+    }
+
+    friend bool operator!=(const CTransactionBase& a, const CTransactionBase& b)
+    {
+        return !(a==b);
+    }
+
+    // Check for negative or overflow output values
+    bool CheckVout(CAmount& nValueOut, CValidationState &state) const;
+    bool CheckOutputsAreStandard(int nHeight, std::string& reason) const;
+    bool CheckOutputsCheckBlockAtHeightOpCode(CValidationState& state) const;
+
+    // Return sum of txouts.
+    virtual CAmount GetValueOut() const;
+
+    //-----------------
+    // pure virtual interfaces 
+    virtual bool IsNull() const = 0;
+
+    // return fee amount
+    virtual CAmount GetFeeAmount(CAmount valueIn) const = 0;
 
+    // Compute tx size
+    virtual unsigned int CalculateSize() const = 0;
+
+    // Compute modified tx size for priority calculation (optionally given tx size)
+    virtual unsigned int CalculateModifiedSize(unsigned int nTxSize=0) const = 0;
+
+    virtual std::string EncodeHex() const = 0;
+    virtual std::string ToString() const = 0;
+
+    virtual void AddToBlock(CBlock* pblock) const = 0;
+    virtual void AddToBlockTemplate(CBlockTemplate* pblocktemplate, CAmount fee, unsigned int sigops) const = 0;
+
+    virtual bool ContextualCheck(CValidationState& state, int nHeight, int dosLevel) const = 0;
+    virtual bool IsStandard(std::string& reason, int nHeight) const = 0;
+    virtual bool CheckFinal(int flags = -1) const = 0;
+    virtual bool IsApplicableToState(CValidationState& state, int nHeight = -1) const = 0;
+
+    virtual void UpdateCoins(CValidationState &state, CCoinsViewCache& view, int nHeight) const = 0;
```

Comment:
On a separate branch (intro_getter_tx_cert) I have:

1. Introduced the const getters for vin, vout and vjoinsplits for both transactions and certificates
2. I have made vin, vout and vjoinsplits privates so that they are accessed via the getter on the whole codebase
3. With this getters I can redefine UpdateCoins to work on TransactionBase rather than Transaction and reuse it for both txs and certs. As you mentioned in your comment, currently checks on vin and vjoinsplits will be skipped for certs since the associated getters will return empty vector.

@albertog78 , please confirm this implementation is in line with your idea. Once you do that, I'll merge the intro_getter_tx_cert branch into current one.